src/test/java/com/google/cloud/pubsublite/flink/ITSourceAndSinkTest.java

```diff
@@ -0,0 +1,378 @@
+        .name("PSL Sink");
+
+    env.fromSource(
+            new PubsubLiteSource<>(sourceSettings().build()),
+            WatermarkStrategy.noWatermarks(),
+            "testPSL")
+        .addSink(new CollectSink());
+
+    JobClient client = env.executeAsync();
+
+    while (CollectSink.values().size() < 100) {
+      if (client.getJobExecutionResult().isCompletedExceptionally()) {
+        client.getJobExecutionResult().get();
+      }
+      Thread.sleep(100);
+    }
+    assertThat(CollectSink.values()).containsExactlyElementsIn(INTEGER_STRINGS);
+    client.cancel().get();
+  }
+
+  @Test
+  public void testSinkWithFailure() throws Exception {
+    // Set up a publisher which will fail once when attempting to publish the 37th message
+    Publisher<MessageMetadata> publisher =
+        spy(PerServerPublisherCache.getOrCreate(sinkSettings().build().getPublisherConfig()));
+    Mockito.doAnswer(
+            inv -> {
+              Message m = inv.getArgument(0);
+              if (m.data().toStringUtf8().equals(INTEGER_STRINGS.get(37))
+                  && staticSet.add("publishFailOnce")) {
+                return ApiFutures.immediateFailedFuture(new RuntimeException("failure"));
+              }
+              return inv.callRealMethod();
+            })
+        .when(publisher)
+        .publish(any());
+    PerServerPublisherCache.getCache().set(sinkSettings().build().getPublisherConfig(), publisher);
+
+    StreamExecutionEnvironment env =
+        StreamExecutionEnvironment.getExecutionEnvironment()
+            .setParallelism(2)
+            .enableCheckpointing(100);
+    env.setRestartStrategy(
+        RestartStrategies.fixedDelayRestart(3, Time.of(10, TimeUnit.MILLISECONDS)));
+
+    env.fromCollection(INTEGER_STRINGS)
+        .addSink(new PubsubLiteSink<>(sinkSettings().build()))
+        .name("PSL Sink");
+
+    env.fromSource(
+            new PubsubLiteSource<>(sourceSettings().build()),
+            WatermarkStrategy.noWatermarks(),
+            "testPSL")
+        .addSink(new CollectSink());
+
+    JobClient client = env.executeAsync();
+
+    while (CollectSink.values().size() < 100) {
+      if (client.getJobExecutionResult().isCompletedExceptionally()) {
+        client.getJobExecutionResult().get();
+      }
+      Thread.sleep(100);
+    }
+    assertThat(CollectSink.values()).containsExactlyElementsIn(INTEGER_STRINGS);
+    client.cancel().get();
+  }
+
+  // A set of static strings for simulating persisted storage in pipelines.
+  private static final Set<String> staticSet = Collections.synchronizedSet(new HashSet<>());
+
+  // A testing sink which stores messages in a static map to prevent them from being lost when
+  // the sink is serialized.
+  private static class CollectSink implements SinkFunction<String>, Serializable {
+    // Note: doesn't store duplicates.
+    private static final Set<String> collector = Collections.synchronizedSet(new HashSet<>());
```

Comment:
nit. Instead, I'd use a private static local HashSet, and add synchronized on the methods

Reply:
I'm going to nack this. It's a bit weird since there are some instance methods and some class methods - I think using synchronized invites someone to hold an object level lock when they should be holding the class lock